dataclass_csv/dataclass_reader.py

```diff
@@ -195,14 +195,14 @@ def _process_row(self, row):
 
             try:
                 transformed_value = field_type(value)
-            except ValueError:
+            except ValueError as e:
                 raise CsvValueError(
                     (
                         f'The field `{field.name}` is defined as {field.type} '
                         f'but received a value of type {type(value)}.'
                     ),
                     line_number=self.reader.line_num,
-                ) from None
+                ) from e
```

Comment:
@AlecBenzer what is your opinion on this one? The reason that I implemented that was that I found that the context error in this case the `ValueError` was not important, I was just interested in the new exception from that point which is the `CsvValueError`. What would you think it is the best practice in this case?

With this change I would have error like this:
```
Traceback (most recent call last):
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 197, in _process_row
    transformed_value = field_type(value)
ValueError: invalid literal for int() with base 10: 'test'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 22, in <module>
    main()
  File "test.py", line 18, in main
    data = list(reader)
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 212, in __next__
    return self._process_row(row)
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 199, in _process_row
    raise CsvValueError(
dataclass_csv.exceptions.CsvValueError: The field `age` is defined as <class 'int'> but received a value of type <class 'str'>. [CSV Line number: 2]
```
With the `from None` it will suppress:
```
Traceback (most recent call last):
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 197, in _process_row
    transformed_value = field_type(value)
ValueError: invalid literal for int() with base 10: 'test'
```

Reply:
Wait, are you sure? I think the suppression works the other way; we'd see only:

```
Traceback (most recent call last):
  File "test.py", line 22, in <module>
    main()
  File "test.py", line 18, in main
    data = list(reader)
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 212, in __next__
    return self._process_row(row)
  File "/home/daniel/Reviews/dataclass-csv/dataclass_csv/dataclass_reader.py", line 199, in _process_row
    raise CsvValueError(
dataclass_csv.exceptions.CsvValueError: The field `age` is defined as <class 'int'> but received a value of type <class 'str'>. [CSV Line number: 2]
```

I think when using the a non-standard type (like `SSN`) getting the specific error from the constructor is useful (which is why I made the change).

For builtin types like `int`, I could see wanting to suppress the `invalid literal for int() with base 10: 'test'` part. But honestly even there, I think the context might be useful in some situations. E.g., `ValueError: invalid literal for int() with base 10: 'l8'` would let the user know they have a typo.

---

But up to you. We can maybe do something fancy like checking if the type is a builtin and altering the exception behavior based on that.

Reply:
Yes, I did a small code here with a dataclass:
```python
@dataclass
class User:
    name: str
    email: str
    age: int
```
Added a test in the `age`field and when I get to that part of the code when raising the error from the `ValueError`, the value error exception is shown in the output.

In my opinion the error returned by the `CsvValueError` describe what happened:
`dataclass_csv.exceptions.CsvValueError: The field `age` is defined as <class 'int'> but received a value of type <class 'str'>. [CSV Line number: 2]`
Perhaps we change that text adding  the value that it was trying to parse, for example:
`dataclass_csv.exceptions.CsvValueError: The field `age` is defined as <class 'int'> but received a value "test" of type <class 'str'>. [CSV Line number: 2]`

Then we would skip showing multiple tracebacks and the CsvValueError would be more descriptive and show the value that was wrong. What do you think?

Reply:
_(Oh I think I just misread what you wrote, I thought you meant "it will suppress, and the result will be:".)_

---

What about for cases with a custom class throwing an error? In those cases I think it's less obvious why a particular string input is invalid, and it's better to not throw away the error information produced by the class's constructor.

If the concern is mostly around not having two tracebacks, we could include just the short description of the initial exception's text in the `CsvValueError` exception.

> In my opinion the error returned by the CsvValueError describe what happened:
`dataclass_csv.exceptions.CsvValueError: The field age is defined as <class 'int'> but received a value of type <class 'str'>. [CSV Line number: 2]`

This does describe what happened but I think it's a little misleading about what the _problem_ actually is. It's totally fine to have a `str` here, as long as that `str` can be parsed as an `int`.

Reply:
Mmmmm yeah, now I see. I think you are absolutely right. I think we can have the raise for the original exception.
And I have tested also in case we want to catch that `CsvValueError`, it works fine.

```python
from dataclasses import dataclass
from dataclass_csv import DataclassReader, CsvValueError


@dataclass
class User:
    name: str
    email: str
    age: int


def main():
    with open("users.csv") as f:
        try:
            reader = DataclassReader(f, User)
            data = list(reader)
            print(data)
        except CsvValueError:
            print("something went wrong!")


if __name__ == "__main__":
    main()
```

I think it is a good improvement in the error.

Reply:
Sounds good! Thanks for bearing with me. :)